rust/main/chains/hyperlane-sealevel/src/application/operation_verifier.rs

```diff
@@ -0,0 +1,144 @@
+    static ref NATIVE_WARP_ROUTES: HashSet<H256> = {
+        HashSet::from([
+            hex_or_base58_to_h256("8DtAGQpcMuD5sG3KdxDy49ydqXUggR1LQtebh2TECbAc").unwrap(),
+            hex_or_base58_to_h256("BXKDfnNkgUNVT5uCfk36sv2GDtK6RwAt9SLbGiKzZkih").unwrap(),
+            hex_or_base58_to_h256("7KD647mgysBeEt6PSrv2XYktkSNLzear124oaMENp8SY").unwrap(),
+            hex_or_base58_to_h256("GPFwRQ5Cw6dTWnmappUKJt76DD8yawxPx28QugfCaGaA").unwrap(),
+        ])
+    };
+}
+
+/// Application operation verifier for Sealevel
+#[derive(new)]
+pub struct SealevelApplicationOperationVerifier {
+    provider: SealevelProvider,
+}
+
+#[async_trait]
+impl ApplicationOperationVerifier for SealevelApplicationOperationVerifier {
+    async fn verify(
+        &self,
+        app_context: &Option<String>,
+        message: &HyperlaneMessage,
+    ) -> Option<ApplicationOperationVerifierReport> {
+        debug!(
+            ?app_context,
+            ?message,
+            "Sealevel application operation verifier",
+        );
+
+        // Separate dependency on self and network invocations for ease of unit testing
+        let check_account_does_not_exist_and_get_minimum = |account: H256| async move {
+            let Ok(false) = self.account_exists(account).await else {
+                return None;
+            };
+
+            self.minimum().await
+        };
+
+        Self::verify_message(
+            app_context,
+            message,
+            check_account_does_not_exist_and_get_minimum,
+        )
+        .await
+    }
+}
+
+impl SealevelApplicationOperationVerifier {
+    async fn verify_message<F, Fut>(
+        app_context: &Option<String>,
+        message: &HyperlaneMessage,
+        check_account_exists_and_get_minimum: F,
+    ) -> Option<ApplicationOperationVerifierReport>
+    where
+        F: FnOnce(H256) -> Fut,
+        Fut: Future<Output = Option<U256>>,
+    {
+        use ApplicationOperationVerifierReport::{AmountBelowMinimum, MalformedMessage};
+
+        Self::verify_context(app_context)?;
+
+        // Starting from this point we assume that we are in a warp route context
+
+        NATIVE_WARP_ROUTES.get(&message.recipient)?;
+
+        let mut reader = Cursor::new(message.body.as_slice());
+        let token_message = match TokenMessage::read_from(&mut reader) {
+            Ok(m) => m,
+            Err(_) => return Some(MalformedMessage(message.clone())),
+        };
+
+        let minimum = check_account_exists_and_get_minimum(token_message.recipient()).await?;
+
+        if token_message.amount() < minimum {
+            return Some(AmountBelowMinimum(minimum, token_message.amount()));
+        }
+
+        None
+    }
+
+    fn verify_context(app_context: &Option<String>) -> Option<()> {
+        let context = match app_context {
+            Some(c) => c,
+            None => return None,
+        };
+
+        if !context.starts_with(WARP_ROUTE_PREFIX) {
+            return None;
+        }
+
+        Some(())
```

Comment:
can probably be even simpler but

```suggestion
        app_context
            .as_ref()
            .map(|context| context.starts_with(WARP_ROUTE_PREFIX))?;
        Some(())
```

Reply:
You are right. It can be done simple:

```suggestion
    app_context
            .as_ref()
            .map(|context| context.starts_with(NATIVE_WARP_ROUTE_PREFIX))?
            .then_some(())
```